File: code/ex113.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    private List<List<Integer>> res = new ArrayList<>();
    
    public List<List<Integer>> pathSum(TreeNode root, int sum) {
        if (root == null)    return res;
        List<Integer> path = new ArrayList<>();
        dfs(root, sum, 0, path);
        return res;
    }
    
    private void dfs(TreeNode root, int sum, int cur, List<Integer> path) {
        cur += root.val;
        path.add(root.val);
        
        if (root.left == null && root.right == null) {
            if (cur == sum)    res.add(new ArrayList<Integer>(path));
            path.remove(path.size() - 1);
            return;
        }
        
        if (root.left != null)    dfs(root.left, sum, cur, path);
        if (root.right != null)    dfs(root.right, sum, cur, path);
        
        path.remove(path.size() - 1);
    }
}

File: code/ex113.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> pathSum(TreeNode* root, int sum) {
        if (root == NULL)    return {};
        vector<vector<int>> res;
        vector<int> cur;
        dfs(root, sum, res, cur);
        return res;
    }
    
private:
    void dfs(TreeNode* root, int sum, vector<vector<int>>& res, vector<int>& cur) {
        if (root->left == NULL && root->right == NULL) {
            if (sum - root->val == 0) {
                cur.push_back(root->val);
                res.push_back(cur);
                cur.pop_back();
            }
            return;
        }
        
        cur.push_back(root->val);
        if (root->left != NULL)    dfs(root->left, sum - root->val, res, cur);
        if (root->right != NULL)    dfs(root->right, sum - root->val, res, cur);
        cur.pop_back();
    }
};